File: python/lmd0_5.py
import math


class Lmd05:
    char_set = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ#&"

    @classmethod
    def lmd0_5(cls, block_length=24) -> str:
        return cls(block_length).solve()

    def __init__(self, block_length):
        # 8  16 24 32
        # 24 48 72 96
        assert block_length % 24 == 0
        assert block_length >= 24
        assert block_length <= 96

        self.byte_array = bytearray()
        self.block_length = block_length
        self.rs = None
        self.cond = None
        self.init_magic_num()

    def set_byte_array(self, byte):
        self.byte_array = bytearray(byte)
        self.init_magic_num()

    def init_magic_num(self):
        self.rs = [0xae3115a1, 0xca0ef573, 0x866b151a, 0x993da81e]
        self.cond = 0

    def bits_transform_1(self, seed) -> int:
        self.rs[0] = ((self.rs[1] & self.rs[2]) | ((~self.rs[1]) & self.rs[3])) ^ self.rs[0]
        self.rs[0] ^= seed
        return self.rs[0]

    def bits_transform_2(self, seed) -> int:
        self.rs[1] = (self.rs[2] | self.rs[0] & ~self.rs[3]) ^ self.rs[1]
        self.rs[1] ^= seed
        return self.rs[1]

    def bits_transform_3(self, seed) -> int:
        self.rs[2] = self.rs[0] ^ self.rs[1] ^ self.rs[2] ^ self.rs[3]
        self.rs[2] ^= seed
        return self.rs[2]

    def bits_transform_4(self, seed) -> int:
        self.rs[3] = (self.rs[1] ^ ~self.rs[3] ^ (~self.rs[0] | self.rs[2]))
        self.rs[3] ^= seed
        return self.rs[3]

    def bits_transform(self, seed, output_bits, input_bits=32) -> int:
        t_full = Lmd05.get_full_int(seed, input_bits)
        self.cond = (self.cond ^ t_full) + 1 & 3

        if self.cond == 0:
            return self.bits_transform_1(seed) & (1 << output_bits) - 1
        elif self.cond == 1:
            return self.bits_transform_2(seed) & (1 << output_bits) - 1
        elif self.cond == 2:
            return self.bits_transform_3(seed) & (1 << output_bits) - 1
        elif self.cond == 3:
            return self.bits_transform_4(seed) & (1 << output_bits) - 1

    @staticmethod
    def get_full_int(integer, length) -> int:
        f_int = 0
        for _ in range(math.ceil(32 / length)):
            f_int <<= length
            f_int |= integer
        f_int &= 4294967295
        return f_int

    @staticmethod
    def loop_shift_left(x, n, length=32) -> int:
        return ((x << n) | (x >> length - n)) & ((1 << length) - 1)

    @staticmethod
    def compose_char(int1, int2) -> int:
        hex_num = int1 >> 4 & 3
        hex_num <<= 2
        hex_num ^= int1 & 15
        hex_num <<= 2
        hex_num ^= int1 >> 6 & 3
        hex_num ^= int2 & 15
        return hex_num

    def solve(self) -> str:
        block_length_half = self.block_length // 2
        vector_list = []
        for _ in range(block_length_half):
            vector_list.append(self.bits_transform(0, 8))
        vector_offset = 0

        full = self.block_length - len(self.byte_array) % self.block_length
        for i in range(full):
            full = self.loop_shift_left(full, 1, 8)
            self.byte_array.append(full)

        for offset_block in range(0, len(self.byte_array), self.block_length):
            for i in range(self.block_length):
                index = (i // 2 + vector_offset) % block_length_half
                b = self.byte_array[offset_block + i]
                vector_list[index] ^= self.bits_transform((i << 8) | b, 8, 16)

                vector_offset += b
                vector_offset %= block_length_half

        result_str_list = []

        random_result = self.rs[0] ^ self.rs[1] ^ self.rs[2] ^ self.rs[3]
        for i in range(len(vector_list)):
            vector_list[i] ^= random_result & 255

        for i in range(0, len(vector_list), 3):
            index0 = (i + vector_offset) % block_length_half
            index1 = (index0 + 1) % block_length_half
            index2 = (index0 + 2) % block_length_half
            result_str_list.insert(0, Lmd05.char_set[Lmd05.compose_char(vector_list[index0], vector_list[index1] >> 4)])
            result_str_list.insert(0, Lmd05.char_set[Lmd05.compose_char(vector_list[index2], vector_list[index1])])

        return "".join(result_str_list)


if __name__ == "__main__":
    lmd05 = Lmd05(24)
    lmd05.set_byte_array("0123456789012345678901230123456789".encode(encoding="utf-8"))
    print(lmd05.solve())
    lmd05.set_byte_array("luern0313".encode(encoding="utf-8"))
    print(lmd05.solve())

File: c++/Lmd05.cpp
using namespace std;

#include "Lmd05.h"
#include "md5.hpp"
#include <iostream>
#include <cassert>
#include <vector>
#include <string>
#include <ctime>

const char char_set[65] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ#&";

Lmd05::Lmd05(int block_length) {
    // 8  16 24 32
    // 24 48 72 96
    assert(block_length % 24 == 0);
    assert(block_length >= 24);
    assert(block_length <= 96);

    this->block_length = block_length;

    init_magic_num();
}

void Lmd05::set_byte_array(const char* byte_array, int size) {
    this->byte_array = vector<unsigned char>(byte_array, byte_array + size);
    init_magic_num();
}

void Lmd05::init_magic_num() {
    this->rs[0] = 0xae3115a1;
    this->rs[1] = 0xca0ef573;
    this->rs[2] = 0x866b151a;
    this->rs[3] = 0x993da81e;
    this->cond = 0;
}

int Lmd05::bits_transform(int seed, int output_bits, int input_bits = 32) {
    int t_full = get_full_int(seed, input_bits);
    cond = (cond ^ t_full) + 1 & 3;

    if (cond == 0)
        return bits_transform_1(seed) & (1 << output_bits) - 1;
    else if (cond == 1)
        return bits_transform_2(seed) & (1 << output_bits) - 1;
    else if (cond == 2)
        return bits_transform_3(seed) & (1 << output_bits) - 1;
    else
        return bits_transform_4(seed) & (1 << output_bits) - 1;
}

int Lmd05::get_full_int(int integer, int length) {
    int f_int = 0;
    for (int i = 0; i < ceil(32 / length); i++) {
        f_int <<= length;
        f_int |= integer;
    }
    f_int &= 4294967295;
    return f_int;
}

int Lmd05::loop_shift_left(int x, int n, int length = 32) {
    if (length == 32)
        return ((x << n) | (x >> (length - n)));
    else
        return ((x << n) | (x >> (length - n))) & ((1 << length) - 1);
}

int Lmd05::compose_char(int int1, int int2) {
    int hex_num = int1 >> 4 & 3;
    hex_num <<= 2;
    hex_num ^= int1 & 15;
    hex_num <<= 2;
    hex_num ^= int1 >> 6 & 3;
    hex_num ^= int2 & 15;
    return hex_num;
}

string Lmd05::solve() {
    int block_length_half = int(block_length / 2);
    vector<int> vector_list;
    for (int i = 0; i < block_length_half; i++)
        vector_list.push_back(bits_transform(0, 8));
    int vector_offset = 0;

    int full_length = block_length - (byte_array.size() % block_length);
    int full = full_length;
    for (int i = 0; i < full_length; i++) {
        full = loop_shift_left(full, 1, 8);
        byte_array.push_back(full);
    }

    for (unsigned int offset_block = 0; offset_block < byte_array.size(); offset_block += block_length) {
        for (int i = 0; i < block_length; i++) {
            int index = (int(i / 2) + vector_offset) % block_length_half;
            int b = byte_array[offset_block + i];
            vector_list[index] ^= bits_transform((i << 8) | b, 8, 16);

            vector_offset += b;
            vector_offset %= block_length_half;
        }
    }

    string result_str_list;

    int random_result = rs[0] ^ rs[1] ^ rs[2] ^ rs[3];
    for (unsigned int i = 0; i < vector_list.size(); i++)
        vector_list[i] ^= random_result & 255;

    for (unsigned int i = 0; i < vector_list.size(); i += 3) {
        int index0 = (i + vector_offset) % block_length_half;
        int index1 = (index0 + 1) % block_length_half;
        int index2 = (index0 + 2) % block_length_half;
        result_str_list.insert(result_str_list.begin(), char_set[compose_char(vector_list[index0], vector_list[index1] >> 4)]);
        result_str_list.insert(result_str_list.begin(), char_set[compose_char(vector_list[index2], vector_list[index1])]);
    }

    return result_str_list;
}

int main()
{
    Lmd05 lmd05(24);
    string encrypt = "0123456789012345678901230123456789";
    lmd05.set_byte_array(encrypt.c_str(), encrypt.size());
    printf("%s\n", lmd05.solve().c_str());
    encrypt = "luern0313";
    lmd05.set_byte_array(encrypt.c_str(), encrypt.size());
    printf("%s\n", lmd05.solve().c_str());
}
